### DISCUSSION TITLE
Cubit meshing a few elements manually / meshing a weird sector region

### DISCUSSION THREAD

**Discussion by GiudGiud** [idaholab/moose#18341] [Q&A Meshing] [answered]
Is there a way in Cubit to mesh a few elements manually?
Do I need to specify just a single volume = 1 element and be careful that the volume shape corresponds to an element?
If I create a few elements manually, should I:
- do this region first and the other volumes will pick up on the nodes I created
- do this region last and use the nodes created by the other regions (how do I retrieve them then?)

Alternatively, does anyone know a scheme that can mesh this, and connect with the hex mesh volumes nearby

Here's a closeup on the volume (orange) I'm trying to mesh. It's an open source model.  

<img width="266" alt="Screen Shot 2021-07-14 at 11 50 30 PM" src="https://user-images.githubusercontent.com/25037151/125736049-f5fb051a-ddc6-4636-9ef0-5208509d4d8a.png">

Additional details on the mesh:
- 2D surfaces revolved using the sweep (0 steps in angle) to create 3D volumes
- 3D volumes meshed directly using the auto scheme

Thanks!
Guillaume

**Comment by GregVernon** [comment, marked as answer]
In your picture, it looks like the orange volume is "pyramidal" in shape... would you be able to share the `.cub` file?  If it is, in fact pyramidal, Cubit doesn't have a built-in method for meshing it with hexes (though you could tetmesh it, then `thex` it into hexes...).  Meshing a pyramid is _the_ great challenge in hex-meshing, look up "Schneider's Pyramid".

One option may be to build a mixed-mesh on that volume - essentially building a pyramid-tet mesh on the orange volume that interfaces with the hex elements on the neighboring volumes.  If your volumes are imprinted & merged, trying doing the following:

```
vol {orange_volume_id} scheme tetmesh
mesh volume {orange_volume_id} 
```

One general comment regarding meshing, not necessarily specific to Cubit is that you should mesh the "most constrained" regions of your model first.  And then mesh the "next most constrained" (which are often the neighbors of the most-constrained regions) and so on.

**Reply by GiudGiud** [reply]
Hello

cub file attached.

I would be fine with a single tet/pyramid at the top of the pyramid and the rest being hexes. 
Is there an easy way to do that or should I just cut the top of the pyramid into one block?
Is there a way to ask for a pyramid mesh cell?

I ll try the test mesh -> hex route now
[fhr_core.zip](https://github.com/idaholab/moose/files/6849974/fhr_core.zip)

Thanks for the advice on the ordering of the mesh. I ll definitely do that once I can get a coarse-ish hex mesh for that volume.

**Reply by GiudGiud** [reply]
One issue with the tet mesh is that the elements are really small, and only get smaller with the THex operation. I tried to set a really large size for the initial mesh. Should I define intervals on the curves instead?

```
    cubit.cmd('volume 8 size 20')
    cubit.cmd('volume 8 scheme Tetmesh')
    cubit.cmd('mesh volume 8')
    cubit.cmd('THex volume 8')
```

**Reply by GregVernon** [reply]
See attached meshed geometry.
[fhr_core_meshed.zip](https://github.com/idaholab/moose/files/6851093/fhr_core_meshed.zip)

Here are the cubit commands that I did:

```
open "fhr_core.cub"

## All volumes, except for the challenge volume, are already meshed
## The surfaces of the challenge volume that are shared with neighboring, meshed volumes, are also meshed (with quads)
## We want to mesh the unmeshed surfaces of the challenge volume with triangles
surface 38 37 8  scheme trimesh geometry approximation angle 15 
Trimesher surface gradation 1.3
Trimesher geometry sizing on
mesh surface 38 37 8 

## Mesh the challenge region with the tetmesh scheme, resulting in a pyramid-to-tet mesh
volume 8  scheme tetmesh
mesh volume 8 
```

In the below image: tet elements are green, pyramid are yellow, hexes are magenta
![image](https://user-images.githubusercontent.com/868479/126392908-7bed11b7-0c2f-40d1-86cf-fedfe16ce513.png)

And detail of the challenge volume:
![image](https://user-images.githubusercontent.com/868479/126393102-df6d1ae5-823f-480c-ba5b-7aab3818197b.png)
![image](https://user-images.githubusercontent.com/868479/126393145-8a10506f-9a13-4a03-823c-2cffc13a8369.png)

The `tetmesh` scheme in Cubit is from the [3D Precise Mesh](https://www.spatial.com/products/3d-precise-mesh) library licensed from Dassault, which they acquired from Distene (library was Distene MeshGems).  The tet meshing library that they provide will _automatically_ produce a mixed tet-pyramid mesh if the input volume has even a single quad element within its bounding surfaces.

## More information
The reason this was challenging _is because_ you have a pyramid-like volume (grey).  In this below picture, you made a webcut (highlighted) that resulted in this volume's shape/topology:
![image](https://user-images.githubusercontent.com/868479/126394050-57cb6e00-9fcf-4ab2-be0d-f7acbca85728.png)

You additionally assigned a block that _only_ contained this _exact_ volume.  This is likely intentional -- you may have a body-load or material assignment that needs to take up _this exact region_.  However, if you didn't mean to have this be a separate block, and it could be united with the blue volume below it, then you could generate a mesh more easily as the resulting body would be prism-like (and thus sweepable):

![image](https://user-images.githubusercontent.com/868479/126394552-b548ca67-9dbc-4af6-bee3-6ab4bc596139.png)

![image](https://user-images.githubusercontent.com/868479/126395102-4acab022-313f-4b4c-a477-76284934db95.png)

Notice how we've effectively relaxed a constraint on the mesh, which makes it easier to hex-mesh.  This may not be a constraint you're able to relax, but it hopefully shows that _sometimes_ "less is more".

**Reply by GiudGiud** [reply]
Thanks for the detailed reply. I was trying to cut the volumes into smaller chunks to make it easier to mesh / narrow down the problem to a single region. We are actually able to relax that constraint.

We are good with this mesh now. Thanks a lot for helping out.
